server/src/main/java/org/apache/druid/server/http/SelfDiscoveryResource.java

```diff
@@ -0,0 +1,85 @@
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.server.http;
+
+import com.google.inject.Inject;
+import com.google.inject.Singleton;
+import com.sun.jersey.spi.container.ResourceFilters;
+import org.apache.druid.discovery.DruidNodeDiscoveryProvider;
+import org.apache.druid.discovery.NodeRole;
+import org.apache.druid.guice.annotations.Self;
+import org.apache.druid.java.util.common.lifecycle.Lifecycle;
+import org.apache.druid.server.DruidNode;
+import org.apache.druid.server.http.security.StateResourceFilter;
+
+import javax.ws.rs.GET;
+import javax.ws.rs.Path;
+import javax.ws.rs.Produces;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+import java.util.Collections;
+import java.util.function.BooleanSupplier;
+
+/**
+ * This class is annotated {@link Singleton} rather than {@link org.apache.druid.guice.LazySingleton} because it adds
+ * a lifecycle handler in the constructor. That should happen before the lifecycle is started, i. e. eagerly during the
+ * DI configuration phase.
+ */
+@Singleton
+@Path("/selfDiscovered")
+@ResourceFilters(StateResourceFilter.class)
+public class SelfDiscoveryResource
+{
+  private BooleanSupplier selfDiscovered;
+
+  @Inject
+  public SelfDiscoveryResource(
+      @Self DruidNode thisDruidNode,
+      @Self NodeRole thisNodeRole,
+      DruidNodeDiscoveryProvider nodeDiscoveryProvider,
+      Lifecycle lifecycle
+  )
+  {
+    Lifecycle.Handler selfDiscoveryListenerRegistrator = new Lifecycle.Handler()
+    {
+      @Override
+      public void start()
+      {
+        selfDiscovered = nodeDiscoveryProvider.getForNode(thisDruidNode, thisNodeRole);
+      }
+
+      @Override
+      public void stop()
+      {
+        // do nothing
+      }
+    };
+    // Using Lifecycle.Stage.SERVER because DruidNodeDiscoveryProvider should be already started when
+    // selfDiscoveryListenerRegistrator.start() is called.
+    lifecycle.addHandler(selfDiscoveryListenerRegistrator, Lifecycle.Stage.SERVER);
+  }
+
+  @GET
+  @Produces(MediaType.APPLICATION_JSON)
+  public Response getSelfDiscovered()
+  {
+    return Response.ok(Collections.singletonMap("selfDiscovered", selfDiscovered.getAsBoolean())).build();
```

Comment:
Maybe better to return 200 OK/503 SERVICE_UNAVAILABLE instead of a JSON response (like `HistoricalResource.getReadiness()`? Some monitoring checks such as AWS load balancer health checks are not able to look at the response body.

Maybe reporting status via response code could be an option.

Reply:
Added `status/selfDiscovered` in addition to `status/selfDiscoveredStatus` which responses in the form of returning 200/503 codes.